pkg/controllers/rollout/controller_integration_test.go

```diff
@@ -0,0 +1,377 @@
+					return false
+				}
+			}
+			return true
+		}, timeout, interval).Should(BeTrue(), "rollout controller should roll all the bindings to Bound state")
+		// simulate that some of the bindings are applied
+		firstApplied := 3
+		for i := 0; i < firstApplied; i++ {
+			markBindingApplied(bindings[i])
+		}
+		// simulate another scheduling decision, pick some cluster to unselect from the bottom of the list
+		var newTarget int32 = 9
+		rolloutCRP.Spec.Policy.NumberOfClusters = &newTarget
+		Expect(k8sClient.Update(ctx, rolloutCRP)).Should(Succeed())
+		secondRoundBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+		deletedBindings := make([]*fleetv1beta1.ClusterResourceBinding, 0)
+		stillScheduled := 6
+		for i := int(targetCluster - 1); i >= stillScheduled; i-- {
+			binding := bindings[i]
+			binding.Spec.State = fleetv1beta1.BindingStateUnscheduled
+			Expect(k8sClient.Update(ctx, binding)).Should(Succeed())
+			By(fmt.Sprintf("resource binding `%s` is marked as not scheduled", binding.Name))
+			deletedBindings = append(deletedBindings, binding)
+		}
+		for i := 0; i < stillScheduled; i++ {
+			secondRoundBindings = append(secondRoundBindings, bindings[i])
+		}
+		// simulate that some of the bindings are applied
+		for i := firstApplied; i < int(newTarget); i++ {
+			markBindingApplied(bindings[i])
+		}
+		newScheduled := int(newTarget) - stillScheduled
+		for i := 0; i < newScheduled; i++ {
+			binding := generateClusterResourceBinding(fleetv1beta1.BindingStateScheduled, masterSnapshot.Name, "cluster-"+strconv.Itoa(int(targetCluster)+i))
+			Expect(k8sClient.Create(ctx, binding)).Should(Succeed())
+			By(fmt.Sprintf("resource binding  %s created", binding.Name))
+			bindings = append(bindings, binding)
+			secondRoundBindings = append(secondRoundBindings, binding)
+		}
+		// simulate that some of the bindings are applied
+		for i := int(newTarget); i < int(targetCluster); i++ {
+			markBindingApplied(bindings[i])
+		}
+		// check that the second round of bindings are scheduled
+		Eventually(func() bool {
+			for _, binding := range secondRoundBindings {
+				err := k8sClient.Get(ctx, types.NamespacedName{Name: binding.GetName()}, binding)
+				if err != nil {
+					return false
+				}
+				if binding.Spec.State != fleetv1beta1.BindingStateBound || binding.Spec.ResourceSnapshotName != masterSnapshot.Name {
+					return false
+				}
+			}
+			return true
+		}, timeout, interval).Should(BeTrue(), "rollout controller should roll all the bindings to Bound state")
+		// simulate that the new bindings are applied
+		for i := 0; i < len(secondRoundBindings); i++ {
+			markBindingApplied(secondRoundBindings[i])
+		}
+		// check that the unselected bindings are deleted
+		Eventually(func() bool {
+			for _, binding := range deletedBindings {
+				err := k8sClient.Get(ctx, types.NamespacedName{Name: binding.GetName()}, binding)
+				if !apierrors.IsNotFound(err) {
+					return false
+				}
+			}
+			return true
```

Comment:
```suggestion
				err := k8sClient.Get(ctx, types.NamespacedName{Name: binding.GetName()}, binding)
				return !apierrors.IsNotFound(err)
```

Reply:
why can we return on the first item in the loop?

Reply:
nvm, i missed the loop there.

nit, could be changed to
```suggestion
				if err := k8sClient.Get(ctx, types.NamespacedName{Name: binding.GetName()}, binding); !apierrors.IsNotFound(err) {
					return false
				}
			}
			return true
```